### PR TITLE
Managing voting power

### PR DESCRIPTION
**Pull request by shivamgarglab49**

### CODE REVIEW

File: contracts/governance/GODHolder.sol [hunk 1]
```diff
@@ -15,37 +15,57 @@ contract GODHolder is TokenHolder {
...
         require(
-            tranferStatus,
+            balance >= _amount,
+            "GODHolder: unlock amount can't be greater to the locked amount"
+        );
+        require(
+            IERC20(_token).transfer(msg.sender, _amount),
             "GODHolder: token transfer failed from contract."
         );
         delete (godTokenForUsers[msg.sender]);
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GODHolder.sol:38]: Why delete here?

`delete (godTokenForUsers[msg.sender]);`

File: e2e-test/business/Governor.ts [hunk 2]
```diff
@@ -442,33 +446,23 @@ export default class Governor extends Base {
...
-      .addAddress(GOD_TOKEN_ID.toSolidityAddress())
-      .addUint256(votingDelay)
-      .addUint256(votingPeriod)
-      .addAddress(this.htsAddress)
-      .addAddress(godHolderProxyAddress)
-      .addUint256(defaultQuorumThresholdValue);
-    await this.execute(900000, INITIALIZE, client, args);
-    console.log(`- Governor#${INITIALIZE}(): done\n`);
-  };
+  ) => {};
```
Comment (amandeeprehal-lab49) [inline e2e-test/business/Governor.ts:449]: Why we need this function now?

File: e2e-test/step-definitions/GovernorTextProposalTest.ts [hunk 3]
```diff
@@ -101,7 +101,7 @@ export class GovernorTextProposal extends CommonSteps {
         console.log(
           `State of proposal is - ${state} revert of god token required is- ${revertRequired}`
         );
-        await godHolder.revertTokensForVoter(clientsInfo.operatorClient);
+        await godHolder.checkAndClaimedGodTokens(clientsInfo.operatorClient);
```
Comment (amandeeprehal-lab49) [inline e2e-test/step-definitions/GovernorTextProposalTest.ts:104]: checkAndClaimedGodTokens -> Why claimed not claim?

I mean this `checkAndClaimGodTokens`

File: integrationTest/governance/governorCreateToken.ts [hunk 4]
```diff
@@ -36,63 +32,21 @@ async function main() {
...
-  await governor.forVote(proposalId, clientsInfo.operatorClient);
-
-  await governor.forVote(proposalId1, clientsInfo.treasureClient);
-  await governor.forVote(proposalId1, clientsInfo.operatorClient);
-
-  await governor.proposalVotes(proposalId);
-  await governor.proposalVotes(proposalId1);
-
+  await governor.getProposalDetails(proposalId);
+  await governor.forVote(proposalId, clientsInfo.uiUserClient);
```
Comment (amandeeprehal-lab49) [inline integrationTest/governance/governorCreateToken.ts:36]: Earlier there were two proposals being used for testing, you are using only one?

File: test/god-holder-test.ts [hunk 5]
```diff
@@ -1,198 +1,246 @@
...
 
-  describe("GODTokenHolderFactory Upgradeable", function () {
-    it("Verify if the GODTokenFactory contract is upgradeable safe ", async function () {
-      const Governor = await ethers.getContractFactory("GODTokenHolderFactory");
-      const args = [zeroAddress, zeroAddress, zeroAddress];
-      const instance = await upgrades.deployProxy(Governor, args);
-      await instance.deployed();
-    });
-  });
+describe("GODHolder tests", function () {
```
Comment (amandeeprehal-lab49) [inline test/god-holder-test.ts:5]: Where is this test. `Verify if the GODTokenFactory contract is upgradeable safe`

File: test/governor-test.ts [hunk 6]
```diff
@@ -0,0 +1,632 @@
+import { expect } from "chai";
+import { Contract } from "ethers";
+import { TestHelper } from "./TestHelper";
+import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+
+describe("Governor Tests", function () {
```
Comment (amandeeprehal-lab49) [inline test/governor-test.ts:7]: Good work! @shivamgarglab49 they looks much better and clean!

File: test/god-holder-test.ts (hunk 5 above)
Comment (shivamgarglab49) [reply test/god-holder-test.ts:5]: We are deploying the logic of GodHolder with initialize method call inside TestHelper.

File: test/governor-test.ts (hunk 6 above)
Comment (shivamgarglab49) [reply test/governor-test.ts:7]: Thanks @amandeeprehal-lab49

File: integrationTest/governance/governorCreateToken.ts (hunk 4 above)
Comment (shivamgarglab49) [reply integrationTest/governance/governorCreateToken.ts:48]: Done

**Review by amandeeprehal-lab49 — APPROVED** [review]

**Pull request merged**